diff --git a/src/game/SpellEffects.cpp b/src/game/SpellEffects.cpp
index f4c8687..77d9632 100644
--- a/src/game/SpellEffects.cpp
+++ b/src/game/SpellEffects.cpp
@@ -2240,6 +2240,13 @@ void Spell::EffectApplyAura(SpellEffectIndex eff_idx)
     if(!unitTarget)
         return;
 
+    // spell Ember Blast (Alar)
+    if(m_spellInfo->Id == 34341)
+    {
+        if(unitTarget->GetTypeId() == TYPEID_UNIT && ((Creature*)unitTarget)->GetEntry() == 19514)
+            return;
+    }             
+
     // ghost spell check, allow apply any auras at player loading in ghost mode (will be cleanup after load)
     if ( (!unitTarget->isAlive() && !(IsDeathOnlySpell(m_spellInfo) || IsDeathPersistentSpell(m_spellInfo))) &&
         (unitTarget->GetTypeId() != TYPEID_PLAYER || !((Player*)unitTarget)->GetSession()->PlayerLoading()) )
