diff --git a/src/game/HomeMovementGenerator.cpp b/src/game/HomeMovementGenerator.cpp
index 0f8d909..639cf89 100644
--- a/src/game/HomeMovementGenerator.cpp
+++ b/src/game/HomeMovementGenerator.cpp
@@ -70,15 +70,22 @@ HomeMovementGenerator<Creature>::Update(Creature &owner, const uint32& time_diff
         owner.AddSplineFlag(SPLINEFLAG_WALKMODE);
 
         // restore orientation of not moving creature at returning to home
-        if (owner.GetDefaultMovementType() == IDLE_MOTION_TYPE)
+        if (CreatureData const* data = sObjectMgr.GetCreatureData(owner.GetGUIDLow()))
         {
             // such a mob might need very exact spawning point, hence relocate to spawn-position
-            if (CreatureData const* data = sObjectMgr.GetCreatureData(owner.GetGUIDLow()))
+            if (data->movementType == IDLE_MOTION_TYPE)
             {
                 owner.SetOrientation(data->orientation);
                 owner.SendHeartBeat(false);
             }
         }
+        else if (owner.IsTemporarySummon())
+        {
+            float fX, fY, fZ, fO;
+            owner.GetSummonPoint(fX, fY, fZ, fO);
+            owner.SetOrientation(fO);
+            owner.SendHeartBeat(false);
+        }
 
         owner.LoadCreatureAddon(true);
         owner.AI()->JustReachedHome();
