diff --git a/scripts/eastern_kingdoms/blackwing_lair/boss_ebonroc.cpp b/scripts/eastern_kingdoms/blackwing_lair/boss_ebonroc.cpp
index d7782fa..ef4be27 100644
--- a/scripts/eastern_kingdoms/blackwing_lair/boss_ebonroc.cpp
+++ b/scripts/eastern_kingdoms/blackwing_lair/boss_ebonroc.cpp
@@ -16,8 +16,8 @@
 
 /* ScriptData
 SDName: Boss_Ebonroc
-SD%Complete: 90
-SDComment: Thrash is missing
+SD%Complete: 100
+SDComment:
 SDCategory: Blackwing Lair
 EndScriptData */
 
@@ -29,7 +29,7 @@ enum
     SPELL_SHADOW_FLAME          = 22539,
     SPELL_WING_BUFFET           = 18500,
     SPELL_SHADOW_OF_EBONROC     = 23340,
-    SPELL_THRASH                = 3391,                     // TODO missing
+    SPELL_THRASH                = 3391,
 };
 
 struct MANGOS_DLL_DECL boss_ebonrocAI : public ScriptedAI
@@ -45,12 +45,14 @@ struct MANGOS_DLL_DECL boss_ebonrocAI : public ScriptedAI
     uint32 m_uiShadowFlameTimer;
     uint32 m_uiWingBuffetTimer;
     uint32 m_uiShadowOfEbonrocTimer;
+    uint32 m_uiTrashTimer;
 
     void Reset()
     {
         m_uiShadowFlameTimer        = 15000;                // These times are probably wrong
         m_uiWingBuffetTimer         = 30000;
         m_uiShadowOfEbonrocTimer    = 45000;
+        m_uiTrashTimer              = urand(7000, 21000);
     }
 
     void Aggro(Unit* pWho)
@@ -80,7 +82,7 @@ struct MANGOS_DLL_DECL boss_ebonrocAI : public ScriptedAI
         if (m_uiShadowFlameTimer < uiDiff)
         {
             if (DoCastSpellIfCan(m_creature, SPELL_SHADOW_FLAME) == CAST_OK)
-                m_uiShadowFlameTimer = urand(12000, 15000);
+                m_uiShadowFlameTimer = urand(15000, 18000);
         }
         else
             m_uiShadowFlameTimer -= uiDiff;
@@ -89,7 +91,7 @@ struct MANGOS_DLL_DECL boss_ebonrocAI : public ScriptedAI
         if (m_uiWingBuffetTimer < uiDiff)
         {
             if (DoCastSpellIfCan(m_creature, SPELL_WING_BUFFET) == CAST_OK)
-                m_uiWingBuffetTimer = 25000;
+                m_uiWingBuffetTimer = urand(30000, 35000);
         }
         else
             m_uiWingBuffetTimer -= uiDiff;
@@ -103,6 +105,15 @@ struct MANGOS_DLL_DECL boss_ebonrocAI : public ScriptedAI
         else
             m_uiShadowOfEbonrocTimer -= uiDiff;
 
+        // Trash Timer
+        if (m_uiTrashTimer < uiDiff)
+        {
+            if (DoCastSpellIfCan(m_creature, SPELL_THRASH) == CAST_OK)
+                m_uiTrashTimer = urand(7000, 21000);
+        }
+        else
+            m_uiTrashTimer -= uiDiff;
+
         DoMeleeAttackIfReady();
     }
 };
