diff --git a/scripts/eastern_kingdoms/blackwing_lair/boss_vaelastrasz.cpp b/scripts/eastern_kingdoms/blackwing_lair/boss_vaelastrasz.cpp
index ece3c8d..c7e0ccd 100644
--- a/scripts/eastern_kingdoms/blackwing_lair/boss_vaelastrasz.cpp
+++ b/scripts/eastern_kingdoms/blackwing_lair/boss_vaelastrasz.cpp
@@ -48,9 +48,6 @@ struct MANGOS_DLL_DECL boss_vaelastraszAI : public ScriptedAI
     {
         m_pInstance = (ScriptedInstance*)pCreature->GetInstanceData();
         Reset();
-
-        // TODO Research what actually is supposed to happen here
-        pCreature->setFaction(35);
     }
 
     ScriptedInstance* m_pInstance;
@@ -69,17 +66,18 @@ struct MANGOS_DLL_DECL boss_vaelastraszAI : public ScriptedAI
 
     void Reset()
     {
-        m_uiPlayerGUID                   = 0;
-        m_uiSpeachTimer                  = 0;
-        m_uiSpeachNum                    = 0;
-        m_uiCleaveTimer                  = 8000;            // These times are probably wrong
-        m_uiFlameBreathTimer             = 11000;
-        m_uiBurningAdrenalineCasterTimer = 15000;
-        m_uiBurningAdrenalineTankTimer   = 45000;
-        m_uiFireNovaTimer                = 5000;
-        m_uiTailSweepTimer               = 20000;
-        m_bHasYelled = false;
-        mbIsDoingSpeach = false;
+        m_uiPlayerGUID                      = 0;
+        m_uiSpeachTimer                     = 0;
+        m_uiSpeachNum                       = 0;
+        m_uiCleaveTimer                     = 8000;            // These times are probably wrong
+        m_uiFlameBreathTimer                = 11000;
+        m_uiBurningAdrenalineCasterTimer    = 15000;
+        m_uiBurningAdrenalineTankTimer      = 45000;
+        m_uiFireNovaTimer                   = 5000;
+        m_uiTailSweepTimer                  = 20000;
+        m_bHasYelled                        = false;
+        mbIsDoingSpeach                     = false;
+        m_creature->setFaction(35);
     }
 
     void BeginSpeach(Unit* target)
@@ -196,21 +194,22 @@ struct MANGOS_DLL_DECL boss_vaelastraszAI : public ScriptedAI
         // Burning Adrenaline Caster Timer
         if (m_uiBurningAdrenalineCasterTimer < uiDiff)
         {
-            Unit* pTarget = NULL;
-
-            // TODO Target Selection must be improved!
-            int i = 0 ;
-            while (i < 3)                                   // max 3 tries to get a random target with power_mana
+            // Select a target with mana-bar
+            std::list<Unit*> lValidTargets;
+            ThreatList const& tList = m_creature->getThreatManager().getThreatList();
+            for (ThreatList::const_iterator iter = tList.begin(); iter != tList.end(); ++iter)
             {
-                ++i;
-                pTarget = m_creature->SelectAttackingTarget(ATTACKING_TARGET_RANDOM, 1);
-                if (pTarget)
-                    if (pTarget->getPowerType() == POWER_MANA)
-                        i=3;
+                Unit* pTempTarget = m_creature->GetMap()->GetUnit((*iter)->getUnitGuid());
+                if (pTempTarget && pTempTarget->getPowerType() == POWER_MANA && pTempTarget != m_creature->getVictim())
+                    lValidTargets.push_back(pTempTarget);
             }
-            if (pTarget)                                    // cast on self (see below)
-                pTarget->CastSpell(pTarget, SPELL_BURNING_ADRENALINE, true);
 
+            if (!lValidTargets.empty())
+            {
+                std::list<Unit*>::const_iterator itr = lValidTargets.begin();
+                advance(itr, urand(0, lValidTargets.size() - 1));
+                (*itr)->CastSpell(*itr, SPELL_BURNING_ADRENALINE, true);
+            }
             m_uiBurningAdrenalineCasterTimer = 15000;
         }
         else
