diff --git a/scripts/outland/hellfire_peninsula.cpp b/scripts/outland/hellfire_peninsula.cpp
index 7b44a54..c09a200 100644
--- a/scripts/outland/hellfire_peninsula.cpp
+++ b/scripts/outland/hellfire_peninsula.cpp
@@ -37,6 +37,7 @@ EndContentData */
 
 #include "precompiled.h"
 #include "escort_ai.h"
+#include "pet_ai.h"
 
 /*######
 ## npc_aeranas
@@ -733,6 +734,41 @@ bool QuestAccept_npc_wounded_blood_elf(Player* pPlayer, Creature* pCreature, con
     return true;
 }
 
+/*######
+## npc_fel_guard_hound
+######*/
+
+enum
+{
+	SPELL_CREATE_POODAD = 37688,
+	NPC_DERANGED_HELBOAR= 16863,
+};
+
+struct MANGOS_DLL_DECL npc_fel_guard_houndAI : public ScriptedPetAI
+{
+    npc_fel_guard_houndAI(Creature* pCreature) : ScriptedPetAI(pCreature) { Reset(); }
+
+    void Reset() { }
+
+    void OwnerKilledUnit(Unit* pVictim)
+    {
+        if (pVictim->GetTypeId() == TYPEID_UNIT && pVictim->GetEntry() == NPC_DERANGED_HELBOAR)
+        {
+            // Distance expected?
+            if (m_creature->IsWithinDistInMap(pVictim, 10.0f))
+            {
+			    if (urand(1, 100) <= 35)
+                    DoCastSpellIfCan(m_creature, SPELL_CREATE_POODAD);
+            }
+        }
+    }
+};
+
+CreatureAI* GetAI_npc_fel_guard_hound(Creature* pCreature)
+{
+    return new npc_fel_guard_houndAI(pCreature);
+}
+
 void AddSC_hellfire_peninsula()
 {
     Script *newscript;
@@ -800,4 +836,9 @@ void AddSC_hellfire_peninsula()
     newscript->GetAI = &GetAI_npc_wounded_blood_elf;
     newscript->pQuestAcceptNPC = &QuestAccept_npc_wounded_blood_elf;
     newscript->RegisterSelf();
+
+    newscript = new Script;
+    newscript->Name = "npc_fel_guard_hound";
+    newscript->GetAI = &GetAI_npc_fel_guard_hound;
+    newscript->RegisterSelf();
 }
