diff --git a/scripts/eastern_kingdoms/blackwing_lair/boss_broodlord_lashlayer.cpp b/scripts/eastern_kingdoms/blackwing_lair/boss_broodlord_lashlayer.cpp
index a644314..fd7962f 100644
--- a/scripts/eastern_kingdoms/blackwing_lair/boss_broodlord_lashlayer.cpp
+++ b/scripts/eastern_kingdoms/blackwing_lair/boss_broodlord_lashlayer.cpp
@@ -32,7 +32,7 @@ enum
     SPELL_CLEAVE                = 26350,
     SPELL_BLAST_WAVE            = 23331,
     SPELL_MORTAL_STRIKE         = 24573,
-    SPELL_KNOCK_AWAY            = 25778
+    SPELL_KNOCK_AWAY            = 18670
 };
 
 struct MANGOS_DLL_DECL boss_broodlordAI : public ScriptedAI
@@ -87,7 +87,7 @@ struct MANGOS_DLL_DECL boss_broodlordAI : public ScriptedAI
         if (m_uiCleaveTimer < uiDiff)
         {
             if (DoCastSpellIfCan(m_creature->getVictim(), SPELL_CLEAVE) == CAST_OK)
-                m_uiCleaveTimer = 7000;
+                m_uiCleaveTimer = urand(3000, 12000);
         }
         else
             m_uiCleaveTimer -= uiDiff;
@@ -112,12 +112,8 @@ struct MANGOS_DLL_DECL boss_broodlordAI : public ScriptedAI
 
         if (m_uiKnockAwayTimer < uiDiff)
         {
-            DoCastSpellIfCan(m_creature->getVictim(), SPELL_KNOCK_AWAY);
-            // Drop 50% aggro - TODO should be scriptedEffect?
-            if (m_creature->getThreatManager().getThreat(m_creature->getVictim()))
-                m_creature->getThreatManager().modifyThreatPercent(m_creature->getVictim(), -50);
-
-            m_uiKnockAwayTimer = urand(15000, 30000);
+            if (DoCastSpellIfCan(m_creature->getVictim(), SPELL_KNOCK_AWAY) == CAST_OK)
+                m_uiKnockAwayTimer = urand(30000, 45000);
         }
         else
             m_uiKnockAwayTimer -= uiDiff;
